File: src/main/java/com/ecommerce/entity/Product.java
package com.ecommerce.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "products")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Product {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @NotBlank(message = "Name is required")
    @Size(max = 200)
    @Column(nullable = false)
    private String name;
    
    @NotBlank(message = "Description is required")
    @Column(nullable = false, columnDefinition = "TEXT")
    private String description;
    
    @NotNull(message = "Price is required")
    @DecimalMin(value = "0.0", message = "Price must be positive")
    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal price;
    
    @NotBlank(message = "Category is required")
    @Column(nullable = false)
    private String category;
    
    @NotBlank(message = "Image is required")
    @Column(nullable = false)
    private String image;
    
    @NotNull(message = "Stock is required")
    @Min(value = 0, message = "Stock cannot be negative")
    @Column(nullable = false)
    private Integer stock;
    
    @DecimalMin(value = "0.0", message = "Discount must be positive")
    @Column(precision = 10, scale = 2)
    private BigDecimal discount;
    
    @Column(nullable = false)
    private Boolean isActive = true;
    
    @CreationTimestamp
    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;
    
    @UpdateTimestamp
    private LocalDateTime updatedAt;
}

File: src/types/product.ts
export interface Category {
  id: number;
  name: string;
  description?: string;
  isActive?: boolean;
}

export class Product {
  id: number;
  name: string;
  description: string;
  price: number;
  category: Category;
  image: string;
  stock: number;
  discount?: number; // Precio con descuento (opcional)

  constructor(
    id: number,
    name: string,
    description: string,
    price: number,
    category: Category,
    image: string,
    stock: number,
    discount?: number
  ) {
    this.id = id;
    this.name = name;
    this.description = description;
    this.price = price;
    this.category = category;
    this.image = image;
    this.stock = stock;
    this.discount = discount;
  }

  getFormattedPrice(): string {
    return `$${this.price.toFixed(2)}`;
  }

  getFormattedDiscountPrice(): string {
    return this.discount ? `$${this.discount.toFixed(2)}` : this.getFormattedPrice();
  }

  getActualPrice(): number {
    return this.discount ? this.discount : this.price;
  }

  hasDiscount(): boolean {
    return this.discount != null && this.discount > 0 && this.discount < this.price;
  }

  getDiscountPercentage(): number {
    if (!this.hasDiscount()) return 0;
    return Math.round(((this.price - this.discount!) / this.price) * 100);
  }
}
